File: DSA/done/49 - groupAnagrams.py
# https://leetcode.com/problems/group-anagrams/

from typing import List

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        hm = {}
        for s in strs:
            cnt = [0] * 26
            for c in s:
                cnt[ord(c) - ord('a')] += 1
            conv = tuple(cnt)
            if conv in hm:
                hm[conv].append(s)
            else:
                hm[conv] = [s]

        return list(hm.values())

File: DSA/done/49 - groupAnagrams.cpp
// https://leetcode.com/problems/group-anagrams/

#include <string>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map<string, vector<string>> hm;
        for (const auto& s : strs) {
            vector<int> count(26, 0);
            for (char c : s) {
                count[c - 'a']++;
            }
            string key{ to_string(count[0]) };
            for (int i{1}; i < 26; i++) {
                key += ',' + to_string(count[i]);
            }
            hm[key].push_back(s);
        }

        vector<vector<string>> result;
        for (const auto& pair : hm) {
            result.push_back(pair.second);
        }
        return result;
    }
};